src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/CreateUserService.kt

```diff
@@ -0,0 +1,134 @@
+    if (user != null) {
+      throw UserExistsException("userId", "User with username ${createUser.email} already exist")
+    }
+
+    if (userRepository.findByEmailAndSourceOrderByUsername(createUser.email).count() >= 1) {
+      throw UserExistsException("email", "Email with username ${createUser.email} already exist")
+    }
+
+   /* This check not required as endpoint is  protected with 'MAINTAIN_OAUTH_USERS' role
+    if (!canMaintainUsers(authenticationFacade.getAuthentication().authorities)) {
+      throw CreateUserException("groupCode", "missing")
+    } */
+    val group = createUser.groupCodes?.let { getInitialGroups(it) }
+
+    user = saveUser(createUser)
+    log.info("External User created: {} , generated user id {}", user.name, user.id)
+    if (group != null) {
+      saveUserGroups(user, group)
+      saveUserRoles(user, group)
+    }
+
+    telemetryClient.trackEvent(
+      "ExternalUserCreated",
+      mapOf(
+        "username" to user.name,
+        "admin" to authenticationFacade.getUsername(),
+        "groups" to group?.map { it.groupCode }.toString(),
+      ),
+      null,
+    )
+
+    return user.id
+  }
+
+  private suspend fun saveUser(createUser: CreateUser): User {
+    val user = User(username = StringUtils.upperCase(createUser.email), source = AuthSource.auth)
+    user.email = createUser.email
+    user.setEnabled(true)
+    user.lastName = createUser.lastName
+    user.setFirstName(createUser.firstName)
+    userRepository.save(user)
+
+    return user
+  }
+  private suspend fun saveUserGroups(user: User, group: Set<Group>) {
+    user.id?.let { userId ->
+      val groupIds = group.mapNotNull { it.groupId }.toSet()
+      val userGroups = groupIds
+        .map {
+          UserGroup(userId, it)
+        }
+        .toList()
+        .toSet()
+
+      userGroupCoroutineRepository.saveAll(userGroups).toList()
+      log.debug("User groups created: {} for user name {}", groupIds, user.name)
+    }
+  }
+
+  private suspend fun saveUserRoles(user: User, group: Set<Group>) {
+    user.id?.let { userId ->
+      val roleIds = group
+        .map { it ->
+          roleRepository.findRolesByGroupCode(it.groupCode)
+            .toList()
+            .mapNotNull { it.id }
+        }
+        .flatten()
+        .toSet()
+
+      val userRoles = roleIds.map { UserRole(userId, it) }
+      userRoleCoroutineRepository.saveAll(userRoles).toList()
+      log.debug("User roles created: {} for user name {}", userRoles.map { it.roleId }.toString(), user.name)
+    }
+  }
+  private suspend fun getInitialGroups(
+    groupCodes: Set<String>,
+  ): Set<Group> {
+    val authUserGroups = userGroupService.getAssignableGroups(authenticationFacade.getUsername(), authenticationFacade.getAuthentication().authorities)
+    val groups = authUserGroups.filter { it.groupCode in groupCodes }.toSet()
+    if ((groups.isEmpty())) {
+      throw CreateUserException("groupCode", "notfound")
```

Comment:
looks like some spare brackets that can be removed

Reply:
removed additional bracket